internal/dispatch/caching/caching.go

```diff
@@ -182,22 +183,28 @@ func (cd *CachingDispatcher) DispatchExpand(ctx context.Context, req *v1.Dispatc
 // DispatchLookup implements dispatch.Lookup interface and does not do any caching yet.
 func (cd *CachingDispatcher) DispatchLookup(ctx context.Context, req *v1.DispatchLookupRequest) (*v1.DispatchLookupResponse, error) {
 	cd.lookupTotalCounter.Inc()
+
 	requestKey := lookupRequestToKey(req)
 	if cachedResultRaw, found := cd.c.Get(requestKey); found {
 		cachedResult := cachedResultRaw.(lookupResultEntry)
 		if req.Metadata.DepthRemaining >= cachedResult.depthRequired {
+			log.Trace().Object("using cached lookup", req).Int("result count", len(cachedResult.result.ResolvedOnrs)).Send()
 			cd.lookupFromCacheCounter.Inc()
 			return cachedResult.result, nil
 		}
 	}
 
 	computed, err := cd.d.DispatchLookup(ctx, req)
 
-	// We only want to cache the result if there was no error
-	if err == nil {
+	// We only want to cache the result if there was no error and nothing was excluded.
+	if err == nil && len(computed.Metadata.LookupExcludedDirect) == 0 && len(computed.Metadata.LookupExcludedTtu) == 0 {
```

Comment:
Where do these actually get set?

Reply:
In the Lookup dispatcher whenever something is skipped:

- https://github.com/authzed/spicedb/pull/268/files#diff-e24eb2e567330120b1257156553fbb962c4bcaf4c44ed5366f2ffeb180af35c3R567
- https://github.com/authzed/spicedb/pull/268/files#diff-e24eb2e567330120b1257156553fbb962c4bcaf4c44ed5366f2ffeb180af35c3R342

They are then cleared when the direct dispatch or TTU has completed:
- https://github.com/authzed/spicedb/pull/268/files#diff-e24eb2e567330120b1257156553fbb962c4bcaf4c44ed5366f2ffeb180af35c3R44
- https://github.com/authzed/spicedb/pull/268/files#diff-e24eb2e567330120b1257156553fbb962c4bcaf4c44ed5366f2ffeb180af35c3R481

This ensures that only "fully" resolved lookups are cached

Reply:
Thanks github, `lookup.go` was collapsed, and even your links didn't expand it.